core/services/workflows/v2/engine_test.go

```diff
@@ -713,6 +716,139 @@ func TestEngine_WASMBinary_With_Config(t *testing.T) {
 	})
 }
 
+func TestSecretsFetcher_Integration(t *testing.T) {
+	cmd := "core/services/workflows/test/wasm/v2/cmd/with_secrets"
+	binaryB := wasmtest.CreateTestBinary(cmd, false, t)
+
+	// Define a custom config to validate against
+	giveName := "Foo"
+	giveNum := int32(42)
+	config := fmt.Appendf(nil, "name: %s\nnumber: %d\n", giveName, giveNum)
+	wasmLogger := logger.NewMockLogger(t)
+	module, err := host.NewModule(&host.ModuleConfig{
+		Logger:         wasmLogger,
+		IsUncompressed: true,
+	}, binaryB)
+	require.NoError(t, err)
+
+	capreg := regmocks.NewCapabilitiesRegistry(t)
+	capreg.EXPECT().LocalNode(matches.AnyContext).Return(newNode(t), nil)
+	mc := v2.MockVaultCapability{
+		Fn: func(ctx context.Context, req capabilities.CapabilityRequest) (capabilities.CapabilityResponse, error) {
+			resp := &vault.GetSecretsResponse{
+				Responses: []*vault.SecretResponse{
+					{
+						Id:        "Foo",
+						Namespace: "Default",
+						Owner:     testWorkflowOwnerA,
+						EncryptedDecryptionKeyShares: []*vault.EncryptedShares{
+							{
+								Shares: []string{"encryptedShare1"},
+							},
+						},
+					},
+				},
+			}
+			pb, innerErr := anypb.New(resp)
+			if innerErr != nil {
+				return capabilities.CapabilityResponse{}, innerErr
+			}
+			return capabilities.CapabilityResponse{
+				Payload: pb,
+			}, nil
+		},
+	}
+	capreg.EXPECT().GetExecutable(matches.AnyContext, vault.CapabilityID).Return(mc, nil)
+
+	billingClient := setupMockBillingClient(t)
+
+	cfg := defaultTestConfig(t)
+	cfg.WorkflowConfig = config
+	cfg.Module = module
+	cfg.CapRegistry = capreg
+	cfg.BillingClient = billingClient
+
+	initDoneCh := make(chan error, 1)
+	subscribedToTriggersCh := make(chan []string, 1)
+	resultReceivedCh := make(chan *sdkpb.ExecutionResult, 1)
+	executionFinishedCh := make(chan string, 1)
+	cfg.Hooks = v2.LifecycleHooks{
+		OnInitialized: func(err error) {
+			initDoneCh <- err
+		},
+		OnSubscribedToTriggers: func(triggerIDs []string) {
+			subscribedToTriggersCh <- triggerIDs
+		},
+		OnExecutionFinished: func(executionID string, _ string) {
+			executionFinishedCh <- executionID
+		},
+		OnResultReceived: func(er *sdkpb.ExecutionResult) {
+			resultReceivedCh <- er
+		},
+	}
+
+	triggerMock := &basictriggermock.BasicCapability{}
+	triggerMock.Trigger = func(ctx context.Context, input *basictrigger.Config) (*basictrigger.Outputs, error) {
+		// Validate that config is as expected during subscription phase
+		require.Equal(t, giveName, input.Name)
+		require.Equal(t, giveNum, input.Number)
+		return &basictrigger.Outputs{CoolOutput: "Hello, "}, nil
+	}
+	wrappedTriggerMock := &registry.CapabilityWrapper{
+		Capability: triggerMock,
+	}
+
+	expectedSecret := "combinedSecret"
+	sf, err := v2.NewSecretsFetcher(
+		cfg.CapRegistry,
+		cfg.Lggr,
+		v2.NewSemaphore[[]*sdkpb.SecretResponse](5),
+		cfg.WorkflowOwner,
+		cfg.WorkflowName.String(),
+		func([]string) (string, error) {
+			return "combinedSecret", nil
+		},
```

Comment:
```suggestion
		func(encryptedKeyShares []string) (string, error) {
			return encryptedKeyShares[0], nil
		},
```

To avoid hiding the value the decryptor was called with.